clusterloader2/pkg/measurement/common/api_availability_measurement.go

```
@@ -0,0 +1,248 @@
+func (a *apiAvailabilityMeasurement) Execute(config *measurement.MeasurementConfig) ([]measurement.Summary, error) {
+
+	SSHToMasterSupported := config.ClusterFramework.GetClusterConfig().SSHToMasterSupported
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+
+	probeFrequency, err := util.GetIntOrDefault(config.Params, "frequency", 1)
+	if err != nil {
+		return nil, err
+	}
+
+	switch action {
+	case "start":
+		if a.isRunning {
+			klog.Infof("%s: measurement already running", a)
+			return nil, nil
+		}
+		return a.start(config, SSHToMasterSupported, probeFrequency)
+	case "gather":
+		a.stop()
+		var summaries []measurement.Summary
+		summaries = a.createClusterAvailabilitySummary(probeFrequency, summaries)
+		if SSHToMasterSupported {
+			summaries = a.createMasterAvailabilitySummary(probeFrequency, summaries)
+		}
+		return summaries, nil
+	default:
+		return nil, fmt.Errorf("unknown action %v", action)
+	}
+}
+
+func (a *apiAvailabilityMeasurement) createClusterAvailabilitySummary(probeFrequency int, summaries []measurement.Summary) []measurement.Summary {
+	a.clusterLevelMetrics.mayBeUpdateMaxConsecutiveFailures()
+	clusterAvailabilityMetricsOutput := a.buildApiAvailabilityMetricsOutput(probeFrequency, a.clusterLevelMetrics, true, "")
+	content, err := util.PrettyPrintJSON(clusterAvailabilityMetricsOutput)
+	if err == nil {
+		summaries = append(summaries, measurement.CreateSummary(apiAvailabilityName, "json", content))
+	}
+	return summaries
+}
+
+func (a *apiAvailabilityMeasurement) createMasterAvailabilitySummary(probeFrequency int, summaries []measurement.Summary) []measurement.Summary {
+	for _, host := range a.hosts {
+		a.hostLevelMetrics[host].mayBeUpdateMaxConsecutiveFailures()
+		hostAvailabilityMetricsOutput := a.buildApiAvailabilityMetricsOutput(probeFrequency, a.hostLevelMetrics[host], false, host)
+		content, err := util.PrettyPrintJSON(hostAvailabilityMetricsOutput)
```

don't ignore the error - if error != nil this function should return an error